<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="date" content="2015-07-30" />

<title>Advanced Phylotranscriptomics Analyses</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%7D%0Apre%20%7B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title">Advanced Phylotranscriptomics Analyses</h1>
<h4 class="date"><em>2015-07-30</em></h4>
</div>


<div id="overview" class="section level2">
<h2>Overview</h2>
<p>In the <a href="Introduction.html">Introduction</a> vignette users learned how to perform the most fundamental computations and visualizations for phylotranscriptomics analyses using the <code>myTAI</code> package. Especially in the <em>Introduction</em> vignette we demonstrated how to perform <code>Phylostratigraphy</code> and <code>Divergence Stratigraphy</code> as well as the construction of <code>PhyloExpressionSets</code> and <code>DivergenceExpressionSets</code> using the <code>MatchMap()</code> function.</p>
<p>In the <a href="Intermediate.html">Intermediate</a> vignette users learned about more detailed analyses and more specialized techniques to investigate the observed phylotranscriptomics patterns (<code>TAI</code>, <code>TDI</code>, <code>RE</code>, etc.).</p>
<p>This vignette aims to provide users with in-depth tutorials on how to retrieve published <code>Phylostratigraphic Maps</code> and <code>Divergence Maps</code>, the integrattion of these data sets into cutomized phylotranscriptomics pipelines, and additional types of phylotranscriptomics analyses focusing on the expression level distributions of all age or divergence categories.</p>
</div>
<div id="biological-sequence-retrieval-with-biomartr" class="section level2">
<h2>Biological Sequence Retrieval with <code>biomartr</code></h2>
<p>Both methods <code>Phylostratigraphy</code> and <code>Divergence Stratigraphy</code> need genome or proteome information for their analyses. This section introduces users to the retrieval of biological sequences stored as <code>*.fasta</code> format to be able to perform <code>Phylostratigraphy</code> and <code>Divergence Stratigraphy</code>.</p>
<p>The <em>classical</em> way to download biological sequences is via the <code>Terminal</code> application:</p>
<pre class="shell"><code>
# download CDS file of A. thaliana
curl ftp://ftp.ensemblgenomes.org/pub/
plants/release-23/fasta/arabidopsis_thaliana/
cds/Arabidopsis_thaliana.TAIR10.23.cds.all.fa.gz 
-o Arabidopsis_thaliana.TAIR10.23.cds.all.fa.gz

# download CDS file of A. lyrata
curl ftp://ftp.ensemblgenomes.org/pub/plants/
release-23/fasta/arabidopsis_lyrata/cds/
Arabidopsis_lyrata.v.1.0.23.cds.all.fa.gz 
-o Arabidopsis_lyrata.v.1.0.23.cds.all.fa.gz

</code></pre>
<p>Alternatively, users can use the Biological Data Retrieval package <a href="https://github.com/HajkD/biomartr">biomartr</a> to download proteomes from the refseq database (see <a href="https://github.com/HajkD/biomartr/blob/master/vignettes/Sequence_Retrieval.Rmd">Sequence Retrieval Vignette</a> for details).</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install biomartr from CRAN</span>
<span class="kw">install.packages</span>(<span class="st">&quot;biomartr&quot;</span>,
                 <span class="dt">repos        =</span> <span class="st">&quot;https://cran.rstudio.com/&quot;</span>,
                 <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># download the proteome of Arabidopsis thaliana from refseq</span>
<span class="co"># and store the corresponding proteome file in '_ncbi_downloads/proteomes'</span>
<span class="kw">getProteome</span>( <span class="dt">db       =</span> <span class="st">&quot;refseq&quot;</span>, 
             <span class="dt">kingdom  =</span> <span class="st">&quot;plant&quot;</span>,
             <span class="dt">organism =</span> <span class="st">&quot;Arabidopsis thaliana&quot;</span>,
             <span class="dt">path     =</span> <span class="kw">file.path</span>(<span class="st">&quot;_ncbi_downloads&quot;</span>,<span class="st">&quot;proteomes&quot;</span>) )</code></pre>
<p>The <code>getProteome()</code> function creates a directory named <code>_ncbi_downloads/proteomes</code> into which the orresponding proteome named <code>Arabidopsis_thaliana_protein.faa.gz</code> is downloaded. The <code>read_proteome()</code> function enables users to work with the proteome as data.table object.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># path to proteome: '_ncbi_downloads/proteomes/Arabidopsis_thaliana_protein.faa.gz'</span>
file_path &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="st">&quot;_ncbi_downloads&quot;</span>,<span class="st">&quot;proteomes&quot;</span>,<span class="st">&quot;Arabidopsis_thaliana_protein.faa.gz&quot;</span>)
<span class="co"># read proteome as data.table object</span>
Ath_proteome &lt;-<span class="st"> </span><span class="kw">read_proteome</span>(file_path, <span class="dt">format =</span> <span class="st">&quot;fasta&quot;</span>)</code></pre>
<p>In case users would like to store the proteome file at a different location, they can specify the <code>path = file.path(&quot;put&quot;,&quot;your&quot;,&quot;path&quot;,&quot;here&quot;)</code> argument.</p>
</div>
<div id="download-published-phylostratigraphic-maps-and-divergence-maps" class="section level2">
<h2>Download Published <code>Phylostratigraphic Maps</code> and <code>Divergence Maps</code></h2>
<p>Performing the <code>Phylostratigraphy</code> and <code>Divergence Stratigraphy</code> algorithms is a time consuming and computationally complex task. However, the corresponding <code>Phylostratigraphic Maps</code> and <code>Divergence Maps</code> need to be computed only once for each organism of interest. Hence, already published <code>Phylostratigraphic Maps</code> and <code>Divergence Maps</code> can be used without any additional computational tasks to combine them with any corresponding expression dataset of interest to capture evolutionary signals with <code>myTAI</code>.</p>
<p><a href="https://github.com/HajkD/published_phylomaps">Here</a> we provide a collection of scripts and documentation of already published <code>Phylostratigraphic Maps</code> and <code>Divergence Maps</code> and step by step instructions on how to retrieve them for the use with <code>myTAI</code>.</p>
<p>Only in cases where users whish to create custom <code>Phylostratigraphic Maps</code> and <code>Divergence Maps</code> based on specialized reference databases or more recent genome version, the <code>Phylostratigraphy</code> and <code>Divergence Stratigraphy</code> algorithms need to be performed.</p>
</div>
<div id="investigating-age-or-divergence-category-specific-expression-level-distributions" class="section level2">
<h2>Investigating Age or Divergence Category Specific Expression Level Distributions</h2>
<p>Gene expression levels are a fundamental aspect of phylotranscriptomics studies. In detail, phylotranscriptomic measures aim to quantify the expression intensity of genes deriving from common age or divergence categories to detect stages of evolutionary constraints. Hence, the gene expression distribution of age or divergence categories as well as their differences within and between stages or categories allow us to investigate the age (PS) or divergence (DS) category specific contribution to the corresponding transcriptome.</p>
<p>For this purpose, the <code>PlotCategoryExpr()</code> aims to visualize the expression level distribution of each phylostratum during each time point or experiment as barplot, dot plot, or violin plot enabling users to quantify the age (PS) or divergence (DS) category specific contribution to the corresponding transcriptome.</p>
<p>This way of visualizing the gene expression distribution of each age (PS) or divergence (DS) category during all developmental stages or experiments allows users to detect specific age or divergence categories contributing significant levels of gene expression to the underlying biological process (transcriptome).</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(myTAI)

<span class="kw">data</span>(PhyloExpressionSetExample)

<span class="co"># category-centered visualization of PS specific expression level distributions (log-scale)</span>
<span class="kw">PlotCategoryExpr</span>(<span class="dt">ExpressionSet =</span> PhyloExpressionSetExample,
                 <span class="dt">legendName    =</span> <span class="st">&quot;PS&quot;</span>,
                 <span class="dt">test.stat     =</span> <span class="ot">TRUE</span>,
                 <span class="dt">type          =</span> <span class="st">&quot;category-centered&quot;</span>,
                 <span class="dt">distr.type    =</span> <span class="st">&quot;boxplot&quot;</span>,
                 <span class="dt">log.expr      =</span> <span class="ot">TRUE</span>)</code></pre>
<pre><code>                  Zygote Quadrant Globular Heart Torpedo Bent  Mature
category-centered &quot;***&quot;  &quot;***&quot;    &quot;***&quot;    &quot;***&quot; &quot;***&quot;   &quot;***&quot; &quot;***&quot;</code></pre>
<p>The resulting boxplot illustrates the log expression levels of each phylostratum during each developmental stage. Additionally, a Kruskal-Wallis Rank Sum Test as well as a Benjamini &amp; Hochberg p-value adjustment for multiple comparisons is performed (<code>test.stat = TRUE</code>) to statistically quantify the differences between expression levels of different age or divergence categories. This type of analysis allows users to detect stages or experiments that show high diviation between age or divergence category contributions to the overall transcriptome or no significant deviations of age or divergence categories, suggesting equal age or divergence category contributions to the overall transcriptome. The corresponding P-values are printed to the console using the following notation:</p>
<ul>
<li><p>’*’ = P-Value <span class="math">\(\leq\)</span> 0.05</p></li>
<li><p>’**’ = P-Value <span class="math">\(\leq\)</span> 0.005</p></li>
<li><p>’***’ = P-Value <span class="math">\(\leq\)</span> 0.0005</p></li>
<li><p>‘n.s.’ = not significant = P-Value &gt; 0.05</p></li>
</ul>
<p>In this case all developmental stages show significant differences in phylostratum specific gene expression.</p>
<p><strong>Please notice that users need to define the <code>legendName</code> argument as <code>PS</code> or <code>DS</code> to specify whether the input <code>ExpressionSet</code> is a <code>PhyloExpressionSet</code> (<code>legendName = 'PS'</code>) or <code>DivergenceExpressionSet</code> (<code>legendName = 'DS'</code>).</strong></p>
<p>Alternatively, users can investigate the differences of gene expression <strong>between</strong> all stages or experiments for <strong>each</strong> age or divergence category by specifying <code>type = 'stage-centered'</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(myTAI)

<span class="kw">data</span>(PhyloExpressionSetExample)

<span class="co"># stage-centered visualization of PS specific expression level distributions (log-scale)</span>
<span class="kw">PlotCategoryExpr</span>(<span class="dt">ExpressionSet =</span> PhyloExpressionSetExample,
                 <span class="dt">legendName    =</span> <span class="st">&quot;PS&quot;</span>,
                 <span class="dt">test.stat     =</span> <span class="ot">TRUE</span>,
                 <span class="dt">type          =</span> <span class="st">&quot;stage-centered&quot;</span>,
                 <span class="dt">distr.type    =</span> <span class="st">&quot;boxplot&quot;</span>,
                 <span class="dt">log.expr      =</span> <span class="ot">TRUE</span>)</code></pre>
<pre><code>               PS1   PS2   PS3    PS4 PS5    PS6 PS7    PS8    PS9    PS10  PS11   PS12 
stage-centered &quot;***&quot; &quot;***&quot; &quot;n.s.&quot; &quot;*&quot; &quot;n.s.&quot; &quot;*&quot; &quot;n.s.&quot; &quot;n.s.&quot; &quot;n.s.&quot; &quot;***&quot; &quot;n.s.&quot; &quot;***&quot;</code></pre>
<p>Here, the Kruskal-Wallis Rank Sum Test (with Benjamini &amp; Hochberg p-value adjustment) quantifies whether or not the gene expression distribution of a single age or divergence category significantly changes throughout development or experiments. This type of analysis allows users to detect specific age or divergence categories that significantly change their expression levels throughout development or experiments.</p>
<p>In this case, users will observe that PS3,5,7-9,11 do not show significant differences of gene expression between developmental stages suggesting that their contribution to the overall transcriptome remains constant throughout development.</p>
<p>Finally, users can choose the following plot types to visualize expression distributions:</p>
<p>Argument: <code>distr.type</code></p>
<ul>
<li><p><code>distr.type = &quot;boxplot&quot;</code> This specification allows users to visualize the expression distribution of all PS or DS as boxplot.</p></li>
<li><p><code>distr.type = &quot;violin&quot;</code> This specification allows users to visualize the expression distribution of all PS or DS as violin plot.</p></li>
<li><p><code>distr.type = &quot;dotplot&quot;</code> This specification allows users to visualize the expression distribution of all PS or DS as dot plot.</p></li>
</ul>
<p>Together, studies perfomed with <code>PlotCategoryExpr()</code> allow users to conclude that genes originating in specific PS or DS contribute significantly more to the overall transcriptome than other genes originating from different PS or DS categories. More specialized analyses such as <code>PlotMeans()</code>, <code>PlotRE()</code>, <code>PlotBarRE()</code>, <code>TAI()</code>, <code>TDI()</code>, etc. will then allow them to study the exact mean expression patterns of these age or divergence categories.</p>
<p>Users will notice that so far all examples shown above specified <code>log.expr = TRUE</code> illustrating boxplots based on log2 expression levels. This way of visualization allows better visual comparisons between age or divergence categories. However, when specifying <code>log.expr = FALSE</code> absolute expression levels will be visualized in the corresponding boxplot.</p>
<p>Alternatively, instead of specifying <code>log.expr = TRUE</code> users can directly pass log2 transformed expression levels to <code>PlotCategoryExpr()</code> via <code>tf(PhyloExpressionSetExample,log2)</code> (when <code>log.expr = FALSE</code>):</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(PhyloExpressionSetExample)

<span class="co"># category-centered visualization of PS specific expression level distributions (log-scale)</span>
<span class="kw">PlotCategoryExpr</span>(<span class="dt">ExpressionSet =</span> <span class="kw">tf</span>(PhyloExpressionSetExample, log2),
                 <span class="dt">legendName    =</span> <span class="st">&quot;PS&quot;</span>,
                 <span class="dt">test.stat     =</span> <span class="ot">TRUE</span>,
                 <span class="dt">type          =</span> <span class="st">&quot;category-centered&quot;</span>,
                 <span class="dt">distr.type    =</span> <span class="st">&quot;boxplot&quot;</span>,
                 <span class="dt">log.expr      =</span> <span class="ot">FALSE</span>)</code></pre>
<pre><code>                  Zygote Quadrant Globular Heart Torpedo Bent  Mature
category-centered &quot;***&quot;  &quot;***&quot;    &quot;***&quot;    &quot;***&quot; &quot;***&quot;   &quot;***&quot; &quot;***&quot; </code></pre>
<p>Or any other expression level transformation, e.g. <code>sqrt</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(PhyloExpressionSetExample)

<span class="co"># category-centered visualization of PS specific expression level distributions (sqrt-scale)</span>
<span class="kw">PlotCategoryExpr</span>(<span class="dt">ExpressionSet =</span> <span class="kw">tf</span>(PhyloExpressionSetExample, sqrt),
                 <span class="dt">legendName    =</span> <span class="st">&quot;PS&quot;</span>,
                 <span class="dt">test.stat     =</span> <span class="ot">TRUE</span>,
                 <span class="dt">type          =</span> <span class="st">&quot;category-centered&quot;</span>,
                 <span class="dt">distr.type    =</span> <span class="st">&quot;boxplot&quot;</span>,
                 <span class="dt">log.expr      =</span> <span class="ot">FALSE</span>)</code></pre>
<pre><code>                  Zygote Quadrant Globular Heart Torpedo Bent  Mature
category-centered &quot;***&quot;  &quot;***&quot;    &quot;***&quot;    &quot;***&quot; &quot;***&quot;   &quot;***&quot; &quot;***&quot; </code></pre>
</div>
<div id="gene-subset-age-or-divergence-category-specific-expression-level-distributions" class="section level2">
<h2>Gene Subset Age or Divergence Category Specific Expression Level Distributions</h2>
<p>In some cases, users wish to visualize the gene expression distributions for a subset of genes in contrast to the entire transcriptome. For this purpose, the <code>gene.set</code> argument allows users to specify the gene ids of a subset of genes that shall be matched in the input <code>ExpressionSet</code> and for which expression level distributions shall be visualized.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(PhyloExpressionSetExample)

<span class="co"># define an example gene subset (500 genes) which</span>
<span class="co"># can be found in the input ExpressionSet</span>
<span class="kw">set.seed</span>(<span class="dv">234</span>)
example.gene.set &lt;-<span class="st"> </span>PhyloExpressionSetExample[<span class="kw">sample</span>(<span class="dv">1</span>:<span class="dv">25260</span>,<span class="dv">500</span>) , <span class="dv">2</span>]

<span class="co"># visualize the gene expression distributions for these 500 genes (category-centered)</span>
<span class="kw">PlotCategoryExpr</span>(<span class="dt">ExpressionSet =</span> PhyloExpressionSetExample,
                 <span class="dt">legendName    =</span> <span class="st">&quot;PS&quot;</span>,
                 <span class="dt">test.stat     =</span> <span class="ot">TRUE</span>,
                 <span class="dt">type          =</span> <span class="st">&quot;category-centered&quot;</span>,
                 <span class="dt">distr.type    =</span> <span class="st">&quot;boxplot&quot;</span>,
                 <span class="dt">log.expr      =</span> <span class="ot">TRUE</span>,
                 <span class="dt">gene.set      =</span> example.gene.set)</code></pre>
<pre><code>                  Zygote Quadrant Globular Heart Torpedo Bent Mature
category-centered &quot;*&quot;    &quot;*&quot;      &quot;*&quot;      &quot;*&quot;   &quot;*&quot;     &quot;*&quot;  &quot;n.s.&quot;</code></pre>
<p>Or analogously <code>stage-centered</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(PhyloExpressionSetExample)

<span class="co"># define an example gene subset (500 genes) which</span>
<span class="co"># can be found in the input ExpressionSet</span>
<span class="kw">set.seed</span>(<span class="dv">234</span>)
example.gene.set &lt;-<span class="st"> </span>PhyloExpressionSetExample[<span class="kw">sample</span>(<span class="dv">1</span>:<span class="dv">25260</span>,<span class="dv">500</span>) , <span class="dv">2</span>]

<span class="co"># visualize the gene expression distributions for these 500 genes (stage-centered)</span>
<span class="kw">PlotCategoryExpr</span>(<span class="dt">ExpressionSet =</span> PhyloExpressionSetExample,
                 <span class="dt">legendName    =</span> <span class="st">&quot;PS&quot;</span>,
                 <span class="dt">test.stat     =</span> <span class="ot">TRUE</span>,
                 <span class="dt">type          =</span> <span class="st">&quot;stage-centered&quot;</span>,
                 <span class="dt">distr.type    =</span> <span class="st">&quot;boxplot&quot;</span>,
                 <span class="dt">log.expr      =</span> <span class="ot">TRUE</span>,
                 <span class="dt">gene.set      =</span> example.gene.set)</code></pre>
<pre><code>               PS1    PS2    PS3    PS4    PS5    PS6    PS7    PS8    PS9    PS10   PS11  
stage-centered &quot;n.s.&quot; &quot;n.s.&quot; &quot;n.s.&quot; &quot;n.s.&quot; &quot;n.s.&quot; &quot;n.s.&quot; &quot;n.s.&quot; &quot;n.s.&quot; &quot;n.s.&quot; &quot;n.s.&quot; &quot;n.s.&quot;
               PS12  
stage-centered &quot;n.s.&quot;</code></pre>
<p>For example, users interested in the enrichment of PS or DS values in <em>D. rerio</em> brain genes (see <a href="Enrichment.html">Enrichment Vignette</a> for details) could also visualize their gene expression distributions throughout development with <code>PlotCategoryExpr()</code> in cases where expression data is available.</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
